<?xml version="1.0"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="control_mode" default="0" />
  <arg name="type" default="octo" />
  <arg name="estimate_mode"  default= "0" />
  <arg name="headless" default="True" />

  <!-- sensors(fusion) special parameters -->
  <param name="/sensor_plugin/altitude/no_height_offset" value="true" />

  ###########  UAV Config  ###########
  # HYDRUS_XI = 17 # 0x11, hydrus type with 8dof overactuated property
  <param name="/uav_info/uav_model" value= "16" />

  ###########  Motor Config  ###########
  <rosparam file="$(find hydrus_xi)/config/MotorInfo.yaml" command="load" />

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" output="screen" >

    ###########  Basic Param  ###########
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2
    <param name="estimator/estimate_mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>
    <param name="estimator/estimate_mode" value= "2" if="$(arg simulation)" />

    # World Pos Control Mode: 0
    # World Vel Control Mode: 2
    # Local Vel Control Mode: 3
    # Attitude Control Mode: 4
    <param name="navigator/xy_control_mode"  value="$(arg control_mode)"/>

    <param name="simulation" value="$(arg simulation)" /> <!-- for sensor plugin -->
    <param name="param_verbose" value= "true" />
    <param name="main_rate" type="double" value="40"/>

    ###########  Sensor Fusion  ###########
    <rosparam file="$(find hydrus_xi)/config/SensorFusion.yaml" command="load" if="$(arg real_machine)"/>
    <rosparam file="$(find hydrus_xi)/config/SensorSimulation.yaml" command="load" if="$(arg simulation)"/>

    ###########  Overactuated PID Control with Transform Control ############
    <param name="control_plugin_name" value="control_plugin/full_over_actuated" />
    <rosparam file="$(find hydrus_xi)/config/$(arg type)/DifferentialFlatnessPidControlConfig.yaml" command="load" />

    ###########  Teleop  ###########
    <rosparam file="$(find hydrus_xi)/config/TeleopNavigationConfig.yaml" command="load" />
  </node>

  ###########  Sensors  ###########
  <include file="$(find aerial_robot_base)/launch/sensors/onboard_sensors.launch"  if="$(arg real_machine)" >
    <arg name="imu_board" value="1" />  <!-- 1: d imu board, 0: kduino board -->
    <arg name="use_alt" value="1" />
    <arg name="use_mocap" value="0" />
    <arg name="use_joy" value="0" />
    <arg name="use_opti" value="0" />
    <arg name="imu_port" value="/dev/flight_controller" />
    <arg name="alt_port" value="/dev/leddar_one" />
  </include >

  ###########  Servo Bridge  ###########
  <node pkg="hydrus" type="dynamixel_bridge_node" name="servo_bridge"  output="screen"  ns="hydrus_xi" if="$(arg real_machine)" >
    <rosparam file="$(find hydrus_xi)/config/$(arg type)/Servo.yaml" command="load" />
  </node>

  ###########  Robot Model  ###########
  <include file="$(find hydrus_xi)/launch/robot_model.launch" >
    <arg name="type" value="$(arg type)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="need_joint_state" value="false" if="$(arg real_machine)"/>
    <arg name="need_joint_state" value="false" if="$(arg simulation)"/>
    <arg name="description_mode" value="gazebo" if="$(arg simulation)"/>
  </include >

  ########## Simulation in Gazebo #########
  <rosparam file="$(find hydrus_xi)/config/$(arg type)/SimulationControl.yaml" command="load"/>
  <include file="$(find aerial_robot_simulation)/launch/simulation.launch" if = "$(arg simulation)" >
    <arg name="model" default="hydrus_xi" />
    <arg name="gui" default="false" if="$(arg headless)" />
    <arg name="headless" default="$(arg headless)" />
  </include>
</launch>
